### ISSUE TITLE
Volunteer Registration

### ISSUE THREAD

**Issue by runbeyond** [spectrumtrail/spectrum#134]
Problem Statement: Facilitating a volunteer registration process on the website would allow us to passively collect volunteer sign ups, promote volunteer sign up (as an easy process) across social media platforms, and make it easier for us to collect and manage volunteer sign ups.

The basics: a form for users to complete the website and way for administrators to access and view this information from the admin dashboard.

**Form Details:**
Name, email, preferred volunteer position:

**Volunteer Positions:**
Parking
Aid Station (Morning-Noon)
Aid Station (Noon-Afternoon)
Finish Line (high fives and awards)
Pizza Kitchen (11AM-2PM)
Course Tear Down (PM)

For volunteer position selection, a drop down menu or a list with radio dial buttons or whatever other solution you find optimal are acceptable.

**Comment by willblake01** [comment]
@ColinW520 @runbeyond Would we want to give volunteers the ability to manage their registration or only add?

**Comment by willblake01** [comment]
Functional Specifications

1. Facilitate a volunteer registration process on the website to passively collect volunteer sign ups, promote volunteer sign up (as an easy process) across social media platforms, and make it easier for us to collect and manage volunteer sign ups. Deadline is #5 on priority list. Use case: A user would like to volunteer time or give back to the Spectrum community.

2. A form for users to complete the website and way for administrators to access and view this information from the admin dashboard. Requirements: For volunteer position selection, a drop down menu or a list with radio dial buttons or whatever other solution you find optimal are acceptable

3. Project Specifications

3.1 Adds route to volunteer form
	3.1.1 App => config => routes.rb 
3.2 Adds model for form inputs
	3.2.1 App => models => Add form_input.rb
3.3 Adds ability to add volunteer info to database
	3.3.1 Migration => Add volunteers table to database
3.4 Adds controller for form actions
	3.4.1 App => controllers => Add volunteers_controller.rb
3.5 Adds views for form to render on front-end
	3.5.1 App => Add volunteers directory => Add _form.html.haml, new.html.haml, 			_header.html.haml

Basic Flow: User is on Spectrum site and follows a link or button to submit a form indicating their desired level of involvement

Exception Flow: User is on a social media platform and discovers the volunteer opportunity on a post or advertisement and follows a link to the Spectrum volunteer form

Post conditions
Spectrum has persistent access to user volunteer wishes stored in the database and can generate new interest in volunteer opportunities

**Comment by runbeyond** [comment]
Disclaimer that completing the volunteer form does not commit a potential volunteer to anything.

**Comment by willblake01** [comment]
@runbeyond @ColinW520 I've got the basic form ready but 2 questions:
 -Any ideas on how we want to link to the form?
 -Colin mentioned having a general field for positions, do we want to list the positions in a card next to the form or just leave them to enter whatever they want or leave blank?

![image](https://user-images.githubusercontent.com/16193879/95687030-c3b85600-0bc6-11eb-9f5a-6866b7aeb5b2.png)

![image](https://user-images.githubusercontent.com/16193879/95687139-653fa780-0bc7-11eb-8e47-72429b80531c.png)

**Comment by runbeyond** [comment]
Can the positions field also be a pick list?

**Comment by willblake01** [comment]
@ColinW520 Did I misunderstand what you meant from our meeting? ^^

I have in my notes to use a general field for positions. Was the idea that they would enter something if they had a position in mind but the admin would ultimately assign the position?

**Comment by willblake01** [comment]
@runbeyond The only thing with a pick list is we have one for availability "AM/PM/All Day" so it would be incongruent if they were able to choose AM and "Course Tear Down". As I understood it this architecture would force them to focus on and choose a time availability and maybe enter a preference for position and the admin would make the final assignment. To kind of solve the problem of nobody picking "parking" or "course tear down" and everyone choosing "aid station" or "pizza kitchen". Colin's answer should make it clear from an engineering perspective.

**Comment by willblake01** [comment]
Should I remove Vagabond from the event list?

**Comment by runbeyond** [comment]
No, thanks. It’s an active event. If we stop hosting it, we can deactivate
the event and archive the associated race(s).

**Comment by willblake01** [comment]
Oh, sorry, I should have been more clear. Should I not have it in the list as an option to volunteer for? I won't touch it as an actual event. I can just leave it in the list if you could use a volunteer for it or if it's just not worth the trouble of removing.

**Comment by runbeyond** [comment]
Got it! Great catch. It does not make sense to include it in the list of
volunteer events.

**Comment by willblake01** [comment]
I think part of the idea also is if we move it to a microservice then someone else putting on a race using the service might have different positions.

**Comment by runbeyond** [comment]
Can you put together something easily that you can show me?

**Comment by willblake01** [comment]
@runbeyond Which of these columns would you find it valuable to have on the admin dashboard volunteer table showing all volunteers? You'll be able to click into a given volunteer if you want to see all fields but listing them all might get a little cluttered for this view and pollute the relevant data. What questions or insights might you be looking for when scanning a view of all volunteers for all events? Here's the table header with all of them:

![image](https://user-images.githubusercontent.com/16193879/96007805-7f19fe00-0e04-11eb-8364-bb78a4677f53.png)

**Comment by runbeyond** [comment]
@William Blake <gwilliamblake@gmail.com> let's move this to an applicable
story. I will write one now. However, I'd argue that this may not be the
highest priority issue for us to work on now. Arguably, we're putting the
cart in front of the horse with this feature as it has no impact on the
most pressing features which would allow someone to register for a
volunteer position.

**Comment by willblake01** [comment]
@runbeyond @ColinW520 I noticed a conflict in architecture that will block me from proceeding with the first step of the epic. Just wanted to get a consensus for how to proceed. From the details of the epic:

Screen Shot 2020-10-14 at 11.24.03 AM

My notes from Colin's suggestions at the meeting:
-General input form field for position (whatever they enter will go into the position column in database)
-Separate data field for time available
  -Drop down allowing AM, PM, and All Day

*TL:DR How are we going to handle the position input?

**Comment by willblake01** [comment]
If we decide to use the input field for positions we could do something like this in a card for the volunteer to plan their entry, and then later down the line have it dynamically render which positions are still available and how many spots are left.

![image](https://user-images.githubusercontent.com/16193879/96031761-6705a700-0e23-11eb-8ef6-8eb3e2081192.png)

**Comment by runbeyond** [comment]
I like this as minimally viable. @ColinW520 what are your thoughts on this story?

**Comment by willblake01** [comment]
I hope the way I'm writing things out isn't irritating or offensive. Just trying to be super verbose since I'm very new to this and I'm trying to understand the process and learn everything as I go.
